src/main/webapp/app/exercises/programming/participate/programming-submission.service.ts

```diff
@@ -652,4 +656,40 @@ export class ProgrammingSubmissionService implements IProgrammingSubmissionServi
             }
         }
     }
+
+    /**
+     * Locks the given submission, exports it, transforms it to base64, and sends it to Orion
+     *
+     * @param exerciseId id of the exercise the submission belongs to
+     * @param submission submission to send to Orion
+     * @param correctionRound correction round
+     */
+    async downloadSubmissionInOrion(exerciseId: number, submission: Submission | 'new', correctionRound = 0) {
+        this.javaBridge.isCloning(true);
+        const submissionId: number =
+            submission === 'new'
+                ? (await this.getProgrammingSubmissionForExerciseForCorrectionRoundWithoutAssessment(exerciseId, true, correctionRound).toPromise()).id!
+                : submission.id!;
+        const exportOptions: RepositoryExportOptions = {
+            exportAllParticipants: false,
+            filterLateSubmissions: false,
+            addParticipantName: false,
+            combineStudentCommits: false,
+            anonymizeStudentCommits: true,
+            normalizeCodeStyle: false,
+            hideStudentNameInZippedFolder: true,
+        };
+        this.lockAndGetProgrammingSubmissionParticipation(submissionId, correctionRound).subscribe((programmingSubmission: ProgrammingSubmission) => {
+            this.repositoryExportService.exportReposByParticipations(exerciseId, [programmingSubmission.participation!.id!], exportOptions).subscribe((res: HttpResponse<Blob>) => {
```

Comment:
I looked into it. I personally don't prefer these chaining either and therefore will convert the calls to await statements using promises. I found the following sources suggesting this is a reasonable solution:
- "In general, an observable can return multiple values over time. An observable from HttpClient always emits a single value and then completes, never to emit again." [from the angular tutorial](https://angular.io/tutorial/toh-pt6#tap-into-the-observable)
- [This SO thread](https://stackoverflow.com/questions/37364973/what-is-the-difference-between-promises-and-observables/37365955), especially [this answer](https://stackoverflow.com/a/44812252/11948496)
- And [this SO answer](https://stackoverflow.com/a/63744894/11948496)

It makes me wonder however why most other functions I saw rather rely on subscribing - if this is the preferred solution in this project it would make sense to follow it for the sake of uniformity, even if it creates these chained calls in some, rather rare occasions like this.

Reply:
After the discussion in the meeting today I learned promises have caused issues with runtime efficiency in the past and I will therefore revert to using subscriptions to avoid this.